### PR TITLE
discovery/kubernetes: attach node labels when the endpoints TargetRef's kind are Node

### LINKED ISSUE #11961: Node labels for endpoint role doesn't seem to work on services without any runtime in the cluster
### What did you do?

Related with https://github.com/prometheus-community/helm-charts/issues/2961

---

Using the prometheus-operator I'm trying to attach node metadata from  a service that points to Kubelet. 

The Kubelet Service:

```yaml
# Manifest
apiVersion: v1
kind: Service
metadata:
  labels:
    app.kubernetes.io/managed-by: prometheus-operator
    app.kubernetes.io/name: kubelet
    k8s-app: kubelet
  name: kps-system-kubelet
  namespace: kube-system
spec:
  clusterIP: None
  clusterIPs:
  - None
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  - IPv6
  ipFamilyPolicy: RequireDualStack
  ports:
  - name: https-metrics
    port: 10250
    protocol: TCP
    targetPort: 10250
  - name: http-metrics
    port: 10255
    protocol: TCP
    targetPort: 10255
  - name: cadvisor
    port: 4194
    protocol: TCP
    targetPort: 4194
  sessionAffinity: None
  type: ClusterIP

# Describe
Name:              kps-system-kubelet
Namespace:         kube-system
Labels:            app.kubernetes.io/managed-by=prometheus-operator
                   app.kubernetes.io/name=kubelet
                   k8s-app=kubelet
Annotations:       <none>
Selector:          <none>
Type:              ClusterIP
IP Family Policy:  RequireDualStack
IP Families:       IPv4,IPv6
IP:                None
IPs:               None
Port:              https-metrics  10250/TCP
TargetPort:        10250/TCP
Endpoints:         10.XXX.XX.XX:10250,10.XXX.XX.XX:10250,10.XXX.XX.XX:10250 + 5 more...
Port:              http-metrics  10255/TCP
TargetPort:        10255/TCP
Endpoints:         10.XXX.XX.XX:10255,10.XXX.XX.XX:10255,10.XXX.XX.XX:10255 + 5 more...
Port:              cadvisor  4194/TCP
TargetPort:        4194/TCP
Endpoints:         10.XXX.XX.XX:4194,10.XXX.XX.XX:4194,10.XXX.XX.XX:4194 + 5 more...
Session Affinity:  None
Events:            <none>
```

The ServiceMonitor:
```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  labels:
    app: kps-kubelet
    app.kubernetes.io/instance: kps
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/part-of: kps
    app.kubernetes.io/version: 44.3.0
    chart: kps-44.3.0
    heritage: Helm
    prometheus-system: "true"
    release: kps-system
  name: kps-system-kubelet
  namespace: kps-system
spec:
  attachMetadata:
    node: true
  endpoints:
  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
    port: https-metrics
    relabelings:
    - action: replace
      sourceLabels:
      - __metrics_path__
      targetLabel: metrics_path
    scheme: https
    tlsConfig:
      caFile: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
      insecureSkipVerify: true
  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
    metricRelabelings:
    - action: drop
      regex: container_cpu_(cfs_throttled_seconds_total|load_average_10s|system_seconds_total|user_seconds_total)
      sourceLabels:
      - __name__
    - action: drop
      regex: container_fs_(io_current|io_time_seconds_total|io_time_weighted_seconds_total|reads_merged_total|sector_reads_total|sector_writes_total|writes_merged_total)
      sourceLabels:
      - __name__
    - action: drop
      regex: container_memory_(mapped_file|swap)
      sourceLabels:
      - __name__
    - action: drop
      regex: container_(file_descriptors|tasks_state|threads_max)
      sourceLabels:
      - __name__
    - action: drop
      regex: container_spec.*
      sourceLabels:
      - __name__
    - action: drop
      regex: .+;
      sourceLabels:
      - id
      - pod
    path: /metrics/cadvisor
    port: https-metrics
    relabelings:
    - action: replace
      sourceLabels:
      - __metrics_path__
      targetLabel: metrics_path
    scheme: https
    tlsConfig:
      caFile: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
      insecureSkipVerify: true
  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
    path: /metrics/probes
    port: https-metrics
    relabelings:
    - action: replace
      sourceLabels:
      - __metrics_path__
      targetLabel: metrics_path
    scheme: https
    tlsConfig:
      caFile: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
      insecureSkipVerify: true
  jobLabel: k8s-app
  namespaceSelector:
    matchNames:
    - kube-system
  selector:
    matchLabels:
      app.kubernetes.io/name: kubelet
      k8s-app: kubelet
```

The resulting Configuration in prometheus:

```yaml
- job_name: serviceMonitor/monitoring/kube-prometheus-stack-kubelet/0
  honor_labels: true
  honor_timestamps: true
  scrape_interval: 1m
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: https
...
  kubernetes_sd_configs:
  - role: endpoints
    kubeconfig_file: ""
    follow_redirects: true
    enable_http2: true
    namespaces:
      own_namespace: false
      names:
      - kube-system
    attach_metadata:
      node: true
```

Although the parameter is present in the Prometheus configuration, node metadata do not get attached. 
image

### What did you expect to see?

Node metadata attached to the scrape target.

### What did you see instead? Under which circumstances?

No other solution found so far.

### System information

Kube-Prometheus-Stack v44.3.0

### Prometheus version

```text
2.42.0
```

### Prometheus configuration file

```yaml
- job_name: serviceMonitor/monitoring/kube-prometheus-stack-kubelet/0
  honor_labels: true
  honor_timestamps: true
  scrape_interval: 1m
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: https
...
  kubernetes_sd_configs:
  - role: endpoints
    kubeconfig_file: ""
    follow_redirects: true
    enable_http2: true
    namespaces:
      own_namespace: false
      names:
      - kube-system
    attach_metadata:
      node: true
```

### Alertmanager version

```text
0.25.0
```

### Alertmanager configuration file

```yaml
not relevant
```

### Logs

```text
No logs related with the issue, not even in debug mode.
```

### PR DESCRIPTION
**Pull request by mickael-carl**
The current implementation of [attach_metadata](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#kubernetes_sd_config) assumes that Node labels can only be retrieved if the nodeName field is set in endpoints(lices), which means that endpoints where the targetRef is a Node (as created for instance by the Prometheus Operator with its `-kubelet-service` flag) will not have Node labels attached.
For instance, the following Endpoints, when discovered with attach_metadata set to true for nodes:
```yaml
apiVersion: v1
kind: Endpoints
metadata:
  name: kubelet
  namespace: kube-system
subsets:
- addresses:
  - ip: 172.18.0.2
    targetRef:
      kind: Node
      name: apps-local-control-plane
```

With the following labels on the Node object:
```yaml
apiVersion: v1
kind: Node
metadata:
  labels:
    beta.kubernetes.io/arch: arm64
    beta.kubernetes.io/os: linux
    kubernetes.io/arch: arm64
    kubernetes.io/hostname: apps-local-control-plane
    kubernetes.io/os: linux
    node-role.kubernetes.io/control-plane: ""
    node.kubernetes.io/exclude-from-external-load-balancers: ""
  name: apps-local-control-plane
```

Currently results in the following labels:
```
__meta_kubernetes_endpoint_address_target_kind="Node"
__meta_kubernetes_endpoint_address_target_name="apps-local-control-plane"
__meta_kubernetes_endpoint_port_name="https-metrics"
__meta_kubernetes_endpoint_port_protocol="TCP"
__meta_kubernetes_endpoint_ready="true"
__meta_kubernetes_endpoints_label_app_kubernetes_io_managed_by="prometheus-operator"
__meta_kubernetes_endpoints_label_app_kubernetes_io_name="kubelet"
__meta_kubernetes_endpoints_label_k8s_app="kubelet"
__meta_kubernetes_endpoints_labelpresent_app_kubernetes_io_managed_by="true"
__meta_kubernetes_endpoints_labelpresent_app_kubernetes_io_name="true"
__meta_kubernetes_endpoints_labelpresent_k8s_app="true"
__meta_kubernetes_endpoints_name="kubelet"
__meta_kubernetes_namespace="kube-system"
__meta_kubernetes_service_label_app_kubernetes_io_managed_by="prometheus-operator"
__meta_kubernetes_service_label_app_kubernetes_io_name="kubelet"
__meta_kubernetes_service_label_k8s_app="kubelet"
__meta_kubernetes_service_labelpresent_app_kubernetes_io_managed_by="true"
__meta_kubernetes_service_labelpresent_app_kubernetes_io_name="true"
__meta_kubernetes_service_labelpresent_k8s_app="true"
__meta_kubernetes_service_name="kubelet"
```

Note the absence of any meta labels matching the node labels. With this PR however, the following become available:
image

This is done by adding labels when the TargetRef is in fact a Node. 

Fixes #11961.
Might also be worth closing #9510 at the same time?

### CODE REVIEW

File: discovery/kubernetes/endpoints.go
Comment (fpetkovski) [inline discovery/kubernetes/endpoints.go:311]: Is there a case where both of these conditions would be true?

File: discovery/kubernetes/kubernetes.go
Comment (fpetkovski) [inline discovery/kubernetes/kubernetes.go:811]: ditto

File: discovery/kubernetes/kubernetes.go
Comment (fpetkovski) [inline discovery/kubernetes/kubernetes.go:828]: ditto

File: discovery/kubernetes/kubernetes.go
Comment (fpetkovski) [inline discovery/kubernetes/kubernetes.go:778]: Is this continue (and the one on line 781) needed?

**Review by fpetkovski — COMMENTED** [review]
Thanks for the fix

File: discovery/kubernetes/endpoints.go
Comment (mickael-carl) [reply discovery/kubernetes/endpoints.go:311]: As far as I know, the expected behaviour is that it can't but I could be explicit here just to be safe

File: discovery/kubernetes/kubernetes.go
Comment (mickael-carl) [reply discovery/kubernetes/kubernetes.go]: Nope, good point. Fixed!

**Review by fpetkovski — APPROVED** [review]
Not an approver, but this lgtm

**Comment by brancz** [comment]
Not sure why CI is not running but after it runs and it's green happy to merge this.

**Comment by mickael-carl** [comment]
@brancz  I've been wondering the same. It did run on the first patch I submitted but not after force-pushing an update somehow.
@fpetkovski thanks for the feedback!

**Comment by mickael-carl** [comment]
@brancz it might have been [this Github incident](https://www.githubstatus.com/incidents/pr3498h3qkfy). I'll force-push to retrigger.

**Pull request merged**
